src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs

```diff
@@ -0,0 +1,1850 @@
+        }
+    }
+
+    internal sealed class CSharpLazyParameterInitializerOperation : LazyParameterInitializerOperation
+    {
+        private readonly CSharpOperationFactory _operationFactory;
+        private readonly BoundNode _value;
+
+        internal CSharpLazyParameterInitializerOperation(CSharpOperationFactory operationFactory, BoundNode value, ImmutableArray<ILocalSymbol> locals, IParameterSymbol parameter, OperationKind kind, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) : base(locals, parameter, kind, semanticModel, syntax, type, constantValue, isImplicit)
+        {
+            _operationFactory = operationFactory;
+            _value = value;
+        }
+
+        protected override IOperation CreateValue()
+        {
+            return SetParentOperation(_operationFactory.Create(_value), this);
+        }
+    }
+
+    internal sealed class CSharpLazyParenthesizedOperation : LazyParenthesizedOperation
+    {
+        private readonly CSharpOperationFactory _operationFactory;
+        private readonly BoundNode _operand;
+
+        internal CSharpLazyParenthesizedOperation(CSharpOperationFactory operationFactory, BoundNode operand, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) : base(semanticModel, syntax, type, constantValue, isImplicit)
+        {
+            _operationFactory = operationFactory;
+            _operand = operand;
+        }
+
+        protected override IOperation CreateOperand()
+        {
+            return SetParentOperation(_operationFactory.Create(_operand), this);
+        }
+    }
+
+    internal sealed class CSharpLazyPropertyInitializerOperation : LazyPropertyInitializerOperation
+    {
+        private readonly CSharpOperationFactory _operationFactory;
+        private readonly BoundNode _value;
+
+        internal CSharpLazyPropertyInitializerOperation(CSharpOperationFactory operationFactory, BoundNode value, ImmutableArray<ILocalSymbol> locals, ImmutableArray<IPropertySymbol> initializedProperties, OperationKind kind, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) : base(locals, initializedProperties, kind, semanticModel, syntax, type, constantValue, isImplicit)
+        {
+            _operationFactory = operationFactory;
+            _value = value;
+        }
+
+        protected override IOperation CreateValue()
+        {
+            return SetParentOperation(_operationFactory.Create(_value), this);
+        }
+    }
+
+    internal sealed class CSharpLazyPropertyReferenceOperation : LazyPropertyReferenceOperation
+    {
+        private readonly CSharpOperationFactory _operationFactory;
+        private readonly IOperation _instanceNonLazy;
```

Comment:
>private readonly IOperation _instanceNonLazy; [](start = 8, length = 45)

This also feels strange. If we really have to handle both cases, it would probably make sense to create more specialized derived classes and also avoid duplicate storage for the same thing across inheritance hierarchy. I think this can be achieved by passing readily available value to the base and have it be stored in the field of the base in the base constructor.